File: DSA/DynamicProgramming/Code/Memoization.cs
// Memoization Example in C#

using System;

class Memoization {
    static int Fib(int n, int[] dp) {
        if (n <= 1) return n;
        if (dp[n] != -1) return dp[n];
        return dp[n] = Fib(n-1, dp) + Fib(n-2, dp);
    }
    static void Main() {
        Console.WriteLine("=== Memoization Example (C#) ===\n");
        int n = 10;
        int[] dp = new int[100];
        for (int i = 0; i < 100; i++) dp[i] = -1;
        Console.WriteLine($"Fibonacci({n}) = {Fib(n, dp)}");
        Console.WriteLine("\nTime Complexity: O(n)");
        Console.WriteLine("Space Complexity: O(n)");
    }
}

File: DSA/DynamicProgramming/Code/Memoization.py
# Memoization Example in Python

def fib(n, dp):
    if n <= 1:
        return n
    if dp[n] != -1:
        return dp[n]
    dp[n] = fib(n-1, dp) + fib(n-2, dp)
    return dp[n]

n = 10
dp = [-1] * 100
print("=== Memoization Example (Python) ===\n")
print(f"Fibonacci({n}) = {fib(n, dp)}")
print("\nTime Complexity: O(n)")
print("Space Complexity: O(n)")
